Setup target

rwhitby opened the pull request:
With these changes, the build instructions simplify to:

1. git clone --recurse https://github.com/kintex-chatter/xc7k325t-blinky-nextpnr.git
2. cd xc7k325t-blinky-nextpnr
3. make BOARD=qmtech setup
4. make BOARD=qmtech all

I will do a separate pull request with a target to install the prerequisite distribution packages for various platforms.

rwhitby committed:
Add prjxray repository as a submodule.

rwhitby committed:
Add setup and clobber targets to automate the complex steps required to set up the build environment.

Assumes that nextpnr-xilinx and prjxray are submodules.

rwhitby committed:
Run nextpnr-xilinx directly from the install location, so that altering the PATH is not required.

rwhitby committed:
Run fasm2frames from the local python environment and xc7frames2bit directly from the install location.

rwhitby committed:
Update the README to reflect the simplified setup.

unbtorsten commented:
Is there a prefix `${XRAY_DIR}` missing for the path in https://github.com/rwhitby/xc7k325t-blinky-nextpnr/pull/1/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R40?

unbtorsten commented:
This line https://github.com/rwhitby/xc7k325t-blinky-nextpnr/pull/1/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R50 and the following lines  make the Makefile specific to the qmtech-package (xc7k325tffg676-1). The README procedure lays out all the steps and makes it more apparent, where this may be adapted to xc7k325tffg900-2 (package used on Digilent Genesys2). I see an issue with hiding this.

unbtorsten commented:
I see a third unresolved issue in line https://github.com/rwhitby/xc7k325t-blinky-nextpnr/pull/1/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R60, which makes the update of the chipdb conditional. @hansfbaier noted at the end of the README procedure that this has to be updated upon changes in nextpnr. Building the entire environment from scratch will not be confronted with this. A user that chooses to update some the repos might end up in confusion, however.

rwhitby commented:
> Is there a prefix `${XRAY_DIR}` missing for the path in https://github.com/rwhitby/xc7k325t-blinky-nextpnr/pull/1/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R40?

I did not find the fasm2frames in the install location, so I had to run it from the build area.

Is there another way to install prjxray which would put fasm2frames into the install location?

rwhitby commented:
> This line https://github.com/rwhitby/xc7k325t-blinky-nextpnr/pull/1/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R50 and the following lines make the Makefile specific to the qmtech-package (xc7k325tffg676-1). The README procedure lays out all the steps and makes it more apparent, where this may be adapted to xc7k325tffg900-2 (package used on Digilent Genesys2). I see an issue with hiding this.

That is a very good point.  Shall the setup target require the BOARD definition and do the right thing for each board?

unbtorsten commented:
> > Is there a prefix `${XRAY_DIR}` missing for the path in https://github.com/rwhitby/xc7k325t-blinky-nextpnr/pull/1/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R40?
> 
> I did not find the fasm2frames in the install location, so I had to run it from the build area.
> 
> Is there another way to install prjxray which would put fasm2frames into the install location?

See lines 27 and 28 in https://github.com/rwhitby/xc7k325t-blinky-nextpnr/commit/a0a6d09d0295ae666be5acc3b8ba27a1128dff8a#diff-bc037111b143ddd9c78ac9112f228a2aa13eb3fd4f37cf9381bdd82280d16454R27
I think this part was lost in the transition from `makeit.sh` to `Makefile`.

rwhitby commented:
> I see a third unresolved issue in line https://github.com/rwhitby/xc7k325t-blinky-nextpnr/pull/1/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R60, which makes the update of the chipdb conditional. @hansfbaier noted at the end of the README procedure that this has to be updated upon changes in nextpnr. Building the entire environment from scratch will not be confronted with this. A user that chooses to update some the repos might end up in confusion, however.

How about we add a "FORCE=1" variable which overrides all these conditionals and makes them unconditional?

Alternatively, the "clobber" target is intended to re-do everything, but I'm not sure if it should remove the .bba and .bin file as they take ages to rebuild.

I'm open to any suggestions ...

rwhitby commented:
> > > Is there a prefix `${XRAY_DIR}` missing for the path in https://github.com/rwhitby/xc7k325t-blinky-nextpnr/pull/1/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R40?
> > 
> > 
> > I did not find the fasm2frames in the install location, so I had to run it from the build area.
> > Is there another way to install prjxray which would put fasm2frames into the install location?
> 
> See lines 27 and 28 in [a0a6d09#diff-bc037111b143ddd9c78ac9112f228a2aa13eb3fd4f37cf9381bdd82280d16454R27](https://github.com/rwhitby/xc7k325t-blinky-nextpnr/commit/a0a6d09d0295ae666be5acc3b8ba27a1128dff8a#diff-bc037111b143ddd9c78ac9112f228a2aa13eb3fd4f37cf9381bdd82280d16454R27) I think this part was lost in the transition from `makeit.sh` to `Makefile`.

I suspect why it was lost is because a source operation in a Makefile doesn't apply to subsequent lines unless you join all the lines together.  The current Makefile has this issue.  I will attempt to fix this and incorporate those lines from makeit.sh in the next version.

unbtorsten commented:
> > [conditional update of the chipdb]
> 
> How about we add a "FORCE=1" variable which overrides all these conditionals and makes them unconditional?

+1
Sounds like a practical solution to me.

unbtorsten commented:
> > [selection of target Kintex package during make]
> 
> Shall the setup target require the BOARD definition and do the right thing for each board?

I think, that this is the central question. Maybe an issue in kintex-chatter/xc7k325t-blinky-nextpnr would allow for deeper discussion of this aspect?

I guess that Makefiles can naturally handle this by means of different targets. However, I am concerned about the clarity of the make-process. For example, prjxray's hierarchical Makefile structure is highly integrated, but not always easy to understand. A list of commands for different Kintex packages may be a lower barrier for those, that would like to add a new Kintex package. Finally, it should be easy to remove them from our forks once upstream is updated. What do you think?

unbtorsten commented:
Thank you for the update! I have a few more thoughts that I will add directly based off of the code.

unbtorsten commented on .gitmodules:
```diff
@@ -1,3 +1,6 @@
 [submodule "nextpnr-xilinx"]
 	path = nextpnr-xilinx
 	url = git@github.com:kintex-chatter/nextpnr-xilinx.git
+[submodule "prjxray"]
```
Why add prjxray as a submodule? I consider it auxiliary set of binaries and data just like openFPGALoader or yosys that one would have to make available. I agree that setup of the toolchain is a pain these days, but see this to be in the scope of F4PGA (formerly known as Symbiflow), cf. https://github.com/kintex-chatter/xc7k325t-blinky-nextpnr/issues/18#issuecomment-1046351733.

rwhitby replied:
The main reason to add prjxray as a submodule is so that XRAY_DIR can be automatically set to a default value based on this know relative location for prjxray.

unbtorsten commented on Makefile:
```diff
@@ -3,9 +3,15 @@ PREFIX ?= ${HOME}/opt
 SHELL = /bin/bash
 PYTHONPATH ?= ${XRAY_DIR}
 
+ifeq ($(shell uname -s),Darwin)
```
Can you add a comment to explain that this extension refers to Mac (Intel?) only, please?

rwhitby replied:
\# This workaround is only required for macOS, because the default clang does not provide libomp.

unbtorsten commented on Makefile:
```diff
@@ -28,19 +34,48 @@ ${PROJECT_NAME}.json: ${PROJECT_NAME}.v
 ${PROJECT_NAME}.frames: ${PROJECT_NAME}.fasm
-	@. "${XRAY_DIR}/utils/environment.sh"
-	fasm2frames --part ${PART} --db-root ${DB_DIR}/kintex7 $< > $@
+	prjxray/env/bin/fasm2frames --part ${PART} --db-root ${DB_DIR}/kintex7 $< > $@
```
Change to `${XRAY_DIR}/env/bin/fasm2frames --part ${PART} --db-root ${DB_DIR}/kintex7 $< > $@` ?

rwhitby replied:
env/bin does not exist in the installed XRAY_DIR location (\~/opt/prjxray) - it only exists in the local build location (./prjxray), so such a change would break the build.
Is there a way to make fasm2frames get installed into the installed XRAY_DIR location (\~/opt/prjxray) ?

rwhitby replied:
I will add XRAY_UTILS_DIR and XRAY_TOOLS_DIR variables. The former points to the local build area, the latter points to the installed location.

unbtorsten replied:
I think we've had different scenarios in mind.. I had been assuming that an independent clone of prjxray be available in `XRAY_DIR=~/opt/prjxray`, i.e. cloning prjxray would not be controlled through this Makefile. This way, we would have both the `XRAY_UTILS_DIR` and `XRAY_TOOLS_DIR` in the same location. Either way, I think we don't need `make install` for our use of prjxray.

I don't know if fasm2frames can get installed to `XRAY_DIR`. In fact, make install seems to only be concerned with
- $INSTALL_DIR/usr/local/bin/xc7frames2bit
- $INSTALL_DIR/usr/local/bin/xc7patch
- $INSTALL_DIR/usr/local/bin/gen_part_base_yaml
- $INSTALL_DIR/usr/local/bin/frame_address_decoder
- $INSTALL_DIR/usr/local/bin/bittool
- $INSTALL_DIR/usr/local/bin/segmatch
- $INSTALL_DIR/usr/local/bin/bitread

fasm2frames, on the other hand, is part of the build environment.

rwhitby replied:
Yes, I wonder why one is installed and one is left in the build area.  If we need to use fasm2frames for this project, then one would expect others to also need to use it.  Or is there some other way that prjxray is intended to be used which does not require fasm2frames to be called independently?

unbtorsten replied:
There is a related discussion from 2020 at https://github.com/f4pga/prjxray/issues/1384 that resulted in an independent repo for `fasm2frames`: https://github.com/SymbiFlow/f4pga-xc-fasm I have not tested the code therein.

I opened an issue for further clarification at https://github.com/SymbiFlow/f4pga-xc-fasm/issues/19

unbtorsten replied:
The location of fasm2frames seems to be the only (and seemingly unresolved) open question at this time. Why don't you send the pull request to update the make process at kintex-chatter, and we file an issue against https://github.com/rwhitby/xc7k325t-blinky-nextpnr/pull/1/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R45 to track and resolve the need for prjxray sources in the future?

rwhitby replied:
Opened https://github.com/kintex-chatter/xc7k325t-blinky-nextpnr/pull/25 with these notes included.

unbtorsten commented on Makefile:
```diff
@@ -28,19 +34,48 @@ ${PROJECT_NAME}.json: ${PROJECT_NAME}.v
+	${NEXTPNR_BUILD_ENV} cmake -S nextpnr-xilinx -B nextpnr-xilinx/build ${NEXTPNR_CMAKE_FLAGS} -DARCH=xilinx -DCMAKE_INSTALL_PREFIX=${NEXTPNR_DIR}
+	make -C nextpnr-xilinx/build -j2 all
+	make -C nextpnr-xilinx/build install
+	if [ ! -f nextpnr-xilinx/xilinx/xc7k325tffg676-1.bba ] ; then \
```
I suggest to use the part name available from `$PART` instead of a static string (e.g. `xc7k325tffg676-1`). Also, I would support adding a flag to force execution of lines 50 through 64.

rwhitby replied:
Fully agree.

rwhitby commented:
Oh, sorry for any confusion. All I did was rebase against upstream changes. I have not made any intentional updates to address the important items you have raised previously yet. Will get to all this on the weekend.

rwhitby committed:
Use PART to distinguish the target part.

rwhitby committed:
Identify the reason for this platform-specific work-around.

rwhitby committed:
Define XRAY_UTILS_DIR and XRAY_TOOLS_DIR to point to the two different locations used to run fasm2frames and xc7frames2bit.

rwhitby committed:
Use PROJECT_NAME instead of blinky.

rwhitby committed:
Add PROG definition for the QMTech board, using Tigard as the default (but easily overridable) cable.

rwhitby committed:
Genesys2 board has a board definition in OpenFPGALoader, which defaults to the correct digilent cable configuration automatically.

rwhitby committed:
Make the generated files specific to the BOARD, so that multiple BOARDs can be supported in a single project directory.

rwhitby commented:
I believe that my latest push now addresses all comments, and adds board-specific filenames so that multiple boards can be supported from the same project directory.

Once there is agreement here, I will submit the same pull request to the kintex-chatter upstream.

unbtorsten commented:
I have a couple more thoughts:
- See my comment on locating prjxray and fasm2frames in the above code review, cf. https://github.com/rwhitby/xc7k325t-blinky-nextpnr/pull/1#discussion_r814056580 -- I can definitely live with a local clone of prjxray. I'm just wondering whether it is really necessary. After all, we are just using the original upstream code.
- Make `PROJECT_NAME` compatible with overwrite by an environment variable, cf. #21

rwhitby committed:
Allow the project name to be easily overridden.
